ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DiskPolicy.hs

```diff
@@ -0,0 +1,67 @@
+module Test.Ouroboros.Storage.LedgerDB.DiskPolicy (tests) where
+
+import           Data.Function ((&))
+import           Data.Word
+import           Data.Time.Clock (DiffTime, secondsToDiffTime)
+
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import           Ouroboros.Consensus.Config.SecurityParam (SecurityParam(..))
+import           Ouroboros.Consensus.Storage.LedgerDB.DiskPolicy (DiskPolicy(..), defaultDiskPolicy, RequestedSnapshotInterval(..))
+import           Test.Ouroboros.Storage.LedgerDB.OrphanArbitrary ()
+
+tests :: TestTree
+tests = testGroup "DiskPolicy/defaultDiskPolicy" [
+      testGroup "onDiskNumSnapshots" [
+          testProperty "should always be equal to 2" prop_onDiskNumSnapshots
+        ]
+    , testGroup "onDiskShouldTakeSnapshot" [
+          testGroup "haven't take a snapshot yet" [
+            testProperty
+              "should take snapshot if it processed at least k blocks"
+              prop_shouldSnapshot_case1
+        ]
+        , testGroup "have previously taken a snapshot" [
+              testProperty
+                "should take snapshot if time since last is greater then explicitly requested interval"
+                prop_shouldSnapshot_case2
+            , testProperty
+                "should take snapshot if time since last is greater then 2 * k if snapshot interval is set to default"
+                prop_shouldSnapshot_case3
+        ]
+      ]
+    ]
+
+prop_shouldSnapshot_case1 :: Word64 -> SecurityParam -> RequestedSnapshotInterval -> Property
+prop_shouldSnapshot_case1 blocksSinceLast securityParam@(SecurityParam k) requestedInterval =
+  shouldSnapshot blocksSinceLast === (blocksSinceLast >= k)
+  where
+    diskPolicy = defaultDiskPolicy securityParam requestedInterval
+    shouldSnapshot = (onDiskShouldTakeSnapshot diskPolicy) Nothing
+
+prop_shouldSnapshot_case2 :: DiffTime -> Word64 -> SecurityParam -> RequestedSnapshotInterval -> Property
+prop_shouldSnapshot_case2 _ _ _ DefaultSnapshotInterval = discard
```

Comment:
Instead of using `discard`, you can create a `newtype` in `.OrphanArbitrary` that is understood as the argument to `RequestedSnapshotInterval`. This property would take that new type (eg `ActualRequestedSnapshotInterval`, `RequestedSnapshotIntervalArg`, `RequestedSnapshotIntervalValue` etc) unwrap it, and then apply the `RequestedSnapshotInterval` constructor.

All told, that'd let you avoid having to write this "degenerate" case. In general, `discard` rarely passes "the smell test"; it usually indicates the generator can be tightened.

PS - The full `Arbitrary RequestedSnapshotInterval` instance could then reuse this new one.

Reply:
Nice, I was looking for something like this, normally (when using hedgehog) I would just have a bunch of functions returning generators and would never call arbitrary functions in test (just the generators + forAll).

This small little trick closes the gap. Thanks @nfrisby !